### PR TITLE
Prevent various XSS attacks

### PR DESCRIPTION
**Pull request by naNuke**
Hello, I am back with the safe links issue. 
As always this is open to discussion :) but together with already existing markupEscaped setting, this is basically a fix for issues #161 and #210, and maybe other im not aware of.

### CODE REVIEW

File: Parsedown.php [hunk 1]
```diff
@@ -75,6 +75,15 @@ function setUrlsLinked($urlsLinked)
 
     protected $urlsLinked = true;
 
+    function setSafeLinksEnabled($safeLinksEnabled)
+    {
+        $this->safeLinksEnabled = $safeLinksEnabled;
```
Comment (hkdobrev) [inline Parsedown.php]: You'd want to declare this property as a `protected` one and not just leave it to be declared in runtime.

Comment (hkdobrev) [inline Parsedown.php]: Sorry, saw it below later.

**Comment by erusev** [comment]
This looks nice. I'd love to hear more opinions on this. I don't feel like an expert on XSS.

**Comment by xPaw** [comment]
I feel like this param should be named as something more generic like `safeAttributes` and escaping every attribute that user can enter (e.g `title`). Also I think checking for `javascript:` in urls seems like a silly idea, and urls should be escaped regardless.

**Comment by erusev** [comment]
@xPaw Thanks for the input. Which characters do u think should be escaped?

**Comment by xPaw** [comment]
Using `htmlspecialchars` should do the job just fine.

**Comment by erusev** [comment]
Thanks! 

> escaping every attribute that user can enter (e.g title)

Is this about attributes of anchor (`<a>`) tags or all tags?

**Comment by xPaw** [comment]
Any attribute that user can change should be escaped. I guess in Parsedown that only applies to links really.

**Comment by erusev** [comment]
What if instead of having an option about it we implementing it as a default behaviour. It shouldn't have too much performance cost.

**Comment by xPaw** [comment]
That actually would make more sense.

**Comment by naNuke** [comment]
We dont need to escape title attribute because of the way how parsedown works, any additional quote simply breaks the regex and link wont render at all.

thing like ```[example](www.example.com "><script>alert('xss')</script>")```
ends up as ```<a href="www.example.com" title="><script>alert('xss')</script>">```

But adding escape on title is easy if there is really need for it.

The check against javascript: in url is to prevent people from writing clickable javascript links which can be used as an XSS attack.

**Comment by erusev** [comment]
@xPaw Would processing attributes through `htmlspecialchars` protect against things like `[link](javascript:alert('xss'))`?

**Comment by xPaw** [comment]
It won't, but simply checking if the string starts with `javascript:` won't fix the problem either.

See http://security.stackexchange.com/a/73751 for some examples as to why.

**Comment by erusev** [comment]
I asked SO for advice at http://stackoverflow.com/q/28129788/200145

**Comment by naNuke** [comment]
Updated with info from @xPaw to whitelist the protocols instead of javascript: check.
Didn't know the pseudo-protocol can be malformed in some browsers ^^
Also escaping title just to be 100% sure.

**Comment by xPaw** [comment]
I don't think having hardcoded schema list in the regex is a way to go. For example reddit has a bigger schema whitelist: https://github.com/reddit/reddit/blob/bb0f5b552c3a52abfac1e2147144eef8f2e7dfab/r2/r2/lib/filters.py#L177

**Comment by naNuke** [comment]
Thanks for input :) added customizable whitelist, but since parsedown is parser meant for everyone I am not sure if it should include all possible schemas by default. Thoughts?

**Comment by hkdobrev** [comment]
Here a few things I think should be done in order Parsedown to be more safe against XSS.

#### HTML

Raw HTML should be disabled in Markdown. You could use `setMarkupEscaped()` in Parsedown.

#### User-generated content

If a developer is getting user-generated content and is putting it inside Markdown documents this should be escaped/validated beforehand. You could also use HTML Purifier or a similar library to actually validate/escape content after the HTML is generated.

#### Attributes

Every attribute should be escaped for special characters, not only links. In Parsedown this would be the `href`, `title`, `src` and `alt` attributes.

#### Schemas:

1. There is no definitive list of schemas, because you can create your own.
2. [According to OWASP](https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet#XSS_Prevention_Rules_Summary) you should whitelist `http://` and `https://` only for security purposes.

---

Don't take this the wrong way, I think this is a step in the right direction.

About what @erusev suggested that this might be the default:

- Attributes escaping could be done by default.
- HTML escaping is already behind an option and of course should stay that way.
- Schema whitelisting definitely should be optional.
- A note on which says Parsedown is not enough for your content to be secure would be good. Users which use Parsedown should be educated to think about their content and which text comes from where.

**Comment by naNuke** [comment]
Is there any other markdown syntax that accepts user content directly for its generated html other than links and images?

**Comment by naNuke** [comment]
I've just finished reading thru the code and added escaping on *hopefully* all parsed attributes.

On another note, at the moment, all of this expects user input to be UTF-8, which may leave some project vulnerable by not catching html entities under different encodings. [Read here for more info](http://blog.astrumfutura.com/2012/03/a-hitchhikers-guide-to-cross-site-scripting-xss-in-php-part-1-how-not-to-use-htmlspecialchars-for-output-escaping/)

I am not quite sure how to solve the encoding issue since that might be out of scope of this fix, also the fact that parsedown is supporting php 5.2 is not helping, but my idea of possible solution is:

- text() method would accept encoding parameter, encode from that encoding to utf-8, go through all the lines, then re-encode back to user encoding and return.

Not aware of any complications this might bring, but I am not that knowledgeable about unicode.

Let me know what you guys think :)

**Comment by hkdobrev** [comment]
:star2:  LGTM :sparkles: :dizzy:

**Comment by hkdobrev** [comment]
@erusev What's up with this?

**Comment by erusev** [comment]
@hkdobrev I have some thoughts about it. Hopefully, I'll have more time next week to write about them.

**Comment by ivantcholakov** [comment]
Maybe borrowing code from CodeIgniter 3 would be the solution. PHP min = 5.2.4
https://github.com/bcit-ci/CodeIgniter/blob/3.0.0/system/core/Security.php#L349

**Comment by adamthehutt** [comment]
Is there any update on this? Anything I can do to help?

**Comment by hkdobrev** [comment]
> Is there any update on this? Anything I can do to help?

:+1:

**Comment by erusev** [comment]
@adamthehutt You could help with creating simple test data - .md / .html pairs that contain XSS input and the expected safe output.

**Comment by ifeltsweet** [comment]
LGTM

**Comment by malas** [comment]
why this is still not merged into?

**Comment by erusev** [comment]
@malas one reason is the lack of tests

**Comment by malas** [comment]
@erusev then is it safe to say that his project is not supported? such a major issue is not addressed for almost a year, does not sound like a ready for production project.

**Comment by erusev** [comment]
@malas Parsedown is a Markdown parser, its responsibility is to parse Markdown, implementing a safe mode would be nice but it has nothing to do with wether the parser is supported or not so no, it would not be safe to say that the project is not supported

**Comment by tillkruss** [comment]
Any updates on this?

**Comment by aidantwoods** [comment]
If I may, I might suggest that encoding output should be the responsibility of the `element` method so that all attributes and content are always escaped in a consistent way (without having to hunt down all instances of where attributes are set, now or in future).

Choosing a safe URL (i.e. protocol whitelisting) should remain the responsibility of the specific link method though.

However, it may be a better approach to delegate this filtering process elsewhere, so that `href` in links and `src` in images are filtered to be safe at a higher level (e.g. so that extensions that completely override the link method still benefit from this). This argument would yield similar benefits on moving encoding to the `element` method.

**Comment by aidantwoods** [comment]
Hi again everyone, I've gone ahead and implemented my own suggestion on top of the work started here (would be great to get Parsedown a safe mode – given how popular it is).

I've rebased this on master and opened a new PR (can't target this one without running into the conflicts I already had to resolve manually).

Please throw everything you've got at https://github.com/erusev/parsedown/pull/495, by trying to get some XSS through. Or come up with some more test cases if you like.

**Pull request closed without merging**